packages/core/src/storages/request_queue_v2.ts

```diff
@@ -196,48 +259,54 @@ export class RequestQueue extends RequestProvider {
             return;
         }
 
-        this._listHeadAndLockPromise ??= this._listHeadAndLock().finally(() => {
-            this._listHeadAndLockPromise = null;
+        this.listHeadAndLockPromise ??= this._listHeadAndLock().finally(() => {
+            this.listHeadAndLockPromise = null;
         });
 
-        await this._listHeadAndLockPromise;
+        await this.listHeadAndLockPromise;
     }
 
     private async _listHeadAndLock(): Promise<void> {
-        const forefront = await this.hasPendingForefrontRequests();
+        const hasPendingForefrontRequests = await this.hasPendingForefrontRequests();
 
         const headData = await this.client.listAndLockHead({
-            limit: Math.min(forefront ? this.assumedForefrontCount : 25, 25),
+            limit: Math.min(hasPendingForefrontRequests ? this.assumedForefrontCount : 25, 25),
             lockSecs: this.requestLockSecs,
         });
 
+        this.queueHasLockedRequests = headData.queueHasLockedRequests;
+
         const headIdBuffer = [];
 
+        const giveUpLock = async (id?: string, uniqueKey?: string) => {
+            if (id === undefined) {
+                return;
+            }
+
+            try {
+                await this.client.deleteRequestLock(id);
+            } catch {
+                this.log.debug('Failed to delete request lock', { id, uniqueKey });
+            }
+        };
+
+        // If we tried to read new forefront requests, but another client appeared in the meantime, we can't be sure we'll only read our requests.
+        // To retain the correct queue ordering, we rollback this head read.
+        if (hasPendingForefrontRequests && headData.hadMultipleClients) {
+            this.log.debug(`Skipping this read - forefront requests may not be fully consistent`);
+            await Promise.all(headData.items.map(({ id, uniqueKey }) => giveUpLock(id, uniqueKey)));
+        }
+
```

Comment:
@barjin I'm pretty sure this is equivalent to the previous version, but please check it.